vllm/lora/punica.py

```diff
@@ -1,206 +1,260 @@
 # Based on code from https://github.com/punica-ai/punica
 
-from typing import Optional
+from typing import Dict, Optional, Tuple
 
 import torch
 
-from vllm import _custom_ops as ops
+from vllm.lora.ops.bgmv_expand import bgmv_expand
+from vllm.lora.ops.bgmv_expand_slice import bgmv_expand_slice
+from vllm.lora.ops.bgmv_shrink import bgmv_shrink
+from vllm.lora.ops.sgmv_expand import sgmv_expand
+from vllm.lora.ops.sgmv_expand_slice import sgmv_expand_slice
+from vllm.lora.ops.sgmv_shrink import sgmv_shrink
 
+_PARAMS_CACHE: Dict[int, Tuple] = {}
 
-def _check_punica_support():
-    if ops.is_custom_op_supported("_punica_C::dispatch_bgmv"):
-        return
 
-    if torch.cuda.get_device_capability() < (8, 0):
-        raise ImportError(
-            "punica LoRA kernels require compute capability >= 8.0")
-    else:
-        raise ImportError(
-            "punica LoRA kernels could not be imported. If you built vLLM "
-            "from source, make sure VLLM_INSTALL_PUNICA_KERNELS=1 env var "
-            "was set.")
+def _compute_params(
+    token_lora_tensor: torch.Tensor
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, int, int, ]:
+    """
+    Get the information required for the sgmv kernel.
+    """
+    pointer = token_lora_tensor.data_ptr()
+    if pointer not in _PARAMS_CACHE:
+        lora_indices_tensor, seq_length_tensor = torch.unique_consecutive(
+            token_lora_tensor, return_counts=True)
+        cum_result = torch.cumsum(seq_length_tensor, dim=0)
+        b_seq_start_tensor = torch.zeros_like(seq_length_tensor)
+        b_seq_start_tensor[1:].copy_(cum_result[:-1])
+        max_length = seq_length_tensor.max().item()
+        batch_size = lora_indices_tensor.size(0)
+        _PARAMS_CACHE[pointer] = (
+            b_seq_start_tensor,
+            seq_length_tensor,
+            lora_indices_tensor,
+            batch_size,
+            max_length,
+        )
+    return _PARAMS_CACHE[pointer]
 
 
-def bgmv(
+def reset_params_cache():
+    """At the beginning of the prefilling stage, we need  clear the
+    cache explicitly
+    """
+    #TODO release gpu memory
+    torch.cuda.empty_cache()
+    _PARAMS_CACHE.clear()
```

Comment:
swap order so that we can release memory after the cache is cleared?

also why do we need to explicitly call empty_cache in the first place?

Reply:
Yes, this is a bug. It should release memory after clearing the cache. fixed already